### ISSUE TITLE
laplace_marginal_factory: silent non-convergence of inner L-BFGS produces biased marginal density and step_size collapse

### ISSUE THREAD

**Issue by junpenglao** [blackjax-devs/blackjax#925]
## Summary

`laplace_marginal_factory` uses `minimize_lbfgs` with `maxiter=30` (default) for the inner optimization that finds `theta*(phi)`. For ill-conditioned inner problems the optimizer silently hits the iteration ceiling without converging, returning a biased marginal density and biased IFT gradient. The caller receives no warning.

## Failure mode

For a hierarchical GP regression model (d=200 latent variables, inner objective condition number κ≈2600):

- 30 L-BFGS iterations are insufficient: `√κ ≈ 51` iterations needed for gradient-descent convergence; L-BFGS worst-case is O(κ/m) ≈ 260 iterations with memory `m=10`
- The suboptimal `theta_bad` produces `log_p_star = log_joint(theta_bad, phi)` that underestimates the true marginal by `0.5 · ‖theta_bad − theta_star‖²_H` nats
- In testing: density bias up to **173 nats** at some phi values vs the exact analytical marginal
- Because the bias is phi-dependent (L-BFGS converges to different degrees at different phi), the apparent curvature of `log p̂(phi|y)` in phi-space inflates **18–54×** vs the true posterior curvature
- Downstream effect on HMC: adapted `step_size` collapses **670×** (from ~0.1 to ~1.5e-4), acceptance → 1.0, chain frozen

The silent failure is the key problem: `solve_theta` discards `result.state.error` (the gradient norm at termination), so the caller has no signal that the inner solve didn't converge.

## Discriminating evidence

Comparing `marginal_logdensity_fn(phi)` (Laplace, maxiter=30) to the exact analytical GP marginal likelihood `log p(y|phi) = −½ log det(K+σ²I) − ½ yᵀ(K+σ²I)⁻¹y` at 10 GT posterior phi samples:

| | maxiter=30 | maxiter=300 | exact |
|---|---|---|---|
| density bias (max) | 173 nats | < 0.01 nats | 0 |
| corr(log_ks, log_ls) from Hessian | −0.81 (wrong sign) | +0.79 ≈ GT +0.754 | +0.75 |
| HMC step_size | 1.5e-4 (collapsed) | ~0.1 (healthy) | — |

## Proposed fix

**1. Convergence check + warning (the minimum fix):**

In `solve_theta` (and transitively `get_theta_star`), after `minimize_lbfgs` returns, check `result.state.error` (the gradient norm at termination, already computed) against a practical threshold. When the optimizer hits `maxiter` without converging to `gtol`, emit a warning via `jax.debug.callback` (works inside JIT/scan):

```python
result, _ = minimize_lbfgs(objective, initial, **optimizer_kwargs)
# result.state.error is the gradient norm at the final iterate
jax.debug.callback(
    lambda err, nit: warnings.warn(
        f"laplace_marginal_factory: inner L-BFGS did not converge "
        f"(grad_norm={err:.2e} after {int(nit)} iterations). "
        f"The marginal density may be biased. "
        f"Consider increasing maxiter (current={maxiter}; "
        f"for a d-dimensional theta with condition number κ, "
        f"try maxiter=max(300, int(κ**0.5 * 3))).",
        UserWarning, stacklevel=2
    ) if err > 1e-3 else None,
    result.state.error, result.state.iter_num,
    ordered=True
)
```

**2. Raise the default maxiter (or make it condition-number-aware):**

The current default `maxiter=30` is calibrated for well-conditioned problems. For d=200 with κ≈2600, `maxiter=300` is the minimum safe value; `maxiter=500` is recommended to cover L-BFGS worst-case O(κ/m). Options:

- **Option A (conservative):** raise default to `maxiter=100` (3× current; handles κ up to ~100)
- **Option B (robust):** expose a `convergence_tol` parameter; warn when `result.state.error > convergence_tol` regardless of maxiter. Document the κ → maxiter scaling clearly.

**3. Document the maxiter–condition-number relationship in docstring:**

Add to `laplace_marginal_factory` docstring:

```
Convergence notes
-----------------
The inner L-BFGS solve converges in O(√κ) iterations for gradient descent
and O(κ / maxcor) iterations worst-case for L-BFGS (where κ is the condition
number of the Hessian at theta* and maxcor is the history size). The default
maxiter=30 is calibrated for well-conditioned problems (κ < ~100). For
ill-conditioned problems (e.g., dense GP posteriors with d > 100, κ > 1000),
increase maxiter accordingly: ``maxiter=max(300, d // 2)`` is a practical
starting point. Non-convergence silently biases the marginal density.
```

## Impact

This affects all users of `laplace_hmc`, `laplace_dhmc`, `laplace_mhmc`, `laplace_dmhmc` on models where the inner objective is ill-conditioned — including any GP regression, latent factor model, or hierarchical model with a large (d > 50), dense latent block and informative data (which raises the condition number of the precision matrix).

## Reproducer

GP regression with NCP parameterization, d=200 GP function values, N=100 observations:
- Inner objective: `-(‖f_raw‖² + ‖y − L(phi)·f_raw‖²/σ²) / 2` — quadratic, condition number κ = λ_max(I + K/σ²) / λ_min(I + K/σ²) ≈ 2600
- With `maxiter=30`: marginal density biased by up to 173 nats; HMC step_size collapses 670×
- With `maxiter=300`: density matches exact marginal to < 0.01 nats; ridge geometry recovered (+0.79 correlation); step_size healthy

## Related

- `blackjax/mcmc/laplace_marginal.py` — `laplace_marginal_factory`, `solve_theta`, `get_theta_star`
- `blackjax/optimizers/lbfgs.py` — `minimize_lbfgs`, `maxiter=30` default
- The `result.state.error` field (gradient norm) is already available in `OptStep.state` — the check is cheap to add
